go/ndt7-client/source/reader.go

```diff
@@ -0,0 +1,48 @@
+package source
+
+import (
+	"strings"
+	"time"
+
+	"github.com/apex/log"
+	"github.com/gorilla/websocket"
+)
+
+// logmeasurement logs a measurement received from the server. We trim the
+// trailing newlines such that the logs are more compact.
+func logmeasurement(data []byte) {
+	log.Infof("%s", strings.TrimRight(string(data), "\n"))
+}
+
+// Reader runs in a background goroutine and processes all incoming
+// websocket messages. Text messages are logged. The Reader will
+// post any error that may occur on the returned channel. It will
+// then close the channel. It will close the channel without any
+// error being posted in case of normal websocket closure.
+func Reader(conn *websocket.Conn) <-chan error {
```

Comment:
source.Reader and sink.Reader are SO CLOSE to being the same code. Is there a good way of unifying them?

Reply:
Sadly, no. After I reorganized the code to make the race go away, they diverged. To be more specific, in the downloader pipeline, the reader is the source of the pipeline, while in the uploader pipeline it is in the middle of the pipeline. It _could_ still be possible to further compress the code by writing a more abstract implementation, but I'd say this is to be decided after looking again into the changed code. So I am going to resolve this convo for now.